packages/connection-storage/src/connection-storage.ts

```diff
@@ -98,11 +107,128 @@ export class ConnectionStorage {
   }
 
-  private static mapStoredConnectionToConnectionInfo(
-    storedConnectionInfo: ConnectionInfo,
+  static async migrateToSafeStorage() {
+    // If user lands on this version of Compass with legacy connections (using storage mixin),
+    // we will ignore those and only migrate connections that have connectionInfo.
+    const connections = (await this.getConnections()).filter(
+      (x) => x.connectionInfo
+    );
+    if (connections.length === 0) {
+      log.info(
+        mongoLogId(1_001_000_200),
+        'Connection Storage',
+        'No connections to migrate'
+      );
+      return;
+    }
+
+    if (
+      connections.filter((x) => 'version' in x && x.version === this.version)
+        .length === connections.length
+    ) {
+      log.info(
+        mongoLogId(1_001_000_205),
+        'Connection Storage',
+        'Connections already migrated'
+      );
+      return;
+    }
+
+    // Get the secrets from keychain. If there're no secrets, we still want to
+    // migrate the connects to a new version.
+    const secrets = await this.getKeytarCredentials();
+
+    // Connections that are not migrated yet or that failed to migrate,
+    // are not expected to have a version property.
+    const unmigratedConnections = connections.filter((x) => !('version' in x));
+    log.info(
+      mongoLogId(1_001_000_206),
+      'Connection Storage',
+      'Migrating connections',
+      {
+        num_connections: unmigratedConnections.length,
+        num_secrets: Object.keys(secrets).length,
+      }
+    );
+
+    let numFailedToMigrate = 0;
+
+    for (const { connectionInfo } of unmigratedConnections) {
+      // We have connectionInfo as we are limiting this migration to
+      // connections that have connectionInfo (not legacy connections)
+      const _id = connectionInfo!.id;
+      const connectionSecrets = secrets[_id];
+
+      try {
+        await this.userData.write(_id, {
+          _id,
+          connectionInfo, // connectionInfo is without secrets as its direclty read from storage
+          connectionSecrets: this.encryptSecrets(connectionSecrets),
+          version: this.version,
+        });
+      } catch (e) {
+        numFailedToMigrate++;
+        log.error(
+          mongoLogId(1_001_000_207),
+          'Connection Storage',
+          'Failed to migrate connection',
+          { message: (e as Error).message, connectionId: _id }
+        );
+      }
+    }
+
+    log.info(
+      mongoLogId(1_001_000_267),
+      'Connection Storage',
+      'Migration complete',
+      {
+        num_saved_connections:
+          unmigratedConnections.length - numFailedToMigrate,
+        num_failed_connections: numFailedToMigrate,
+      }
+    );
+
+    if (numFailedToMigrate > 0) {
+      track('Keytar Secrets Migration Failed', {
+        num_saved_connections:
+          unmigratedConnections.length - numFailedToMigrate,
+        num_failed_connections: numFailedToMigrate,
+      });
+    }
```

Comment:
Nothin critical, just wanted to point out that in case of a few failed connections we will be tracking both `Migration complete` and `Keytar Secrets Migration Failed`. I wonder if it would be fine just to track `Keytar Secrets Migration Failed`? I don't know the tracking requirements so pardon me if I did not take that into account 😄

Reply:
Sorry I just re-read the highlighted code, one is tracking and one is logging. Please ignore.